### ISSUE TITLE
[FEATURE]Custom node

### ISSUE THREAD

**Issue by LiZhouYan** [FlowiseAI/Flowise#519]
I hope to create a tool for creating nodes that allows users to customize and create nodes directly on the webpage.

**Comment by sam1am** [comment]
This would be great especially if we could share custom nodes back to the project so we could leverage the community more to add support for things.

**Comment by HenryHengZJ** [comment]
what are some of the example custom nodes?

**Comment by LiZhouYan** [comment]
For example, if you define the top-level basic rules of node components, the internal structure of node components can be assembled by dragging and dropping, and specific logic can be executed through custom js or python languages, so that there is no need to customize and develop various components

**Comment by chungyau97** [comment]
Hi @LiZhouYan, 

Do you mean some thing similar to `if...else...` or `for loop` logic node?

**Comment by sam1am** [comment]
I have been looking for a way to concatenate results from several chains. I might be missing a more obvious way I can already do that, but if not then a custom concatenate node could be useful.

**Comment by raffareis** [comment]
Yeah, we definetly could use the ability to add custom logic to the flow. 
We could add decision trees that don't need llms, loggers, transformers...

**Comment by chungyau97** [comment]
Hi @raffareis, do you mean something similar to [n8n switch](https://docs.n8n.io/integrations/builtin/core-nodes/n8n-nodes-base.switch/#node-reference)?

**Comment by raffareis** [comment]
I believe @LiZhouYan meant that. 

I meant something simpler like a Custom Tool, that instead needing an agent to be consumed, could have an input and expose an output as a connection. This node could be used directly on the diagram to act like a "Man-in-the-middle" between nodes, to be able to add logs, tranformations, logic decisions or anything else.

The logic could be added as typescript, just as the custom tool, however the output would have to be configured also, and would need to be exposed as a connection.

One example use for this:

![image](https://github.com/FlowiseAI/Flowise/assets/4365623/1b47226b-610a-452e-8ddd-d5979b80acd8)

**Comment by chungyau97** [comment]
Hey @raffareis, Thanks for the clarification. I get an idea of what you meant.

Below are my understanding on what you guys wanna achieve, correct me if I'm wrong:
**Custom tool node** will be based on an input to make decisions/transform data/perform any action defined by user's written logic, and then output the result to the next node.

**Comment by thebrahman** [comment]
I would like a custom LLM node that gives the ability to use a self hosted LLM by specifying the API endpoint

**Comment by chick26** [comment]
I need this function too. Are there any developer document can help me to rebuild or build a new tool quickly?

**Comment by slavakurilyak** [comment]
+1 for custom nodes as it can unlock custom agents

**Comment by AJV009** [comment]
I REALLY WANT THE CUSTOM NODES!
It would be so cool, only if it was my expertise in the first place, to do something about this.
And Yes, export nodes YES, export flows YES! 
It would be so cool. Flowise would simply go crazy with these little little cool features!
I kinda had assumed before using flowise that it might had some easy way to share and develop custom nodes, because I had a requirement while researching, I want to use something like this flowise in a more varied environment that chat.

**Comment by sidedwards** [comment]
Langflow has the ability to edit the code directly on nodes, which is super helpful.

**Comment by AnhQuanTrl** [comment]
The ability to add CustomNode is a must have I think. There are always a couple of use case that a custom cannot achieve (For example custom MCP client that interact with internal MCP server that has special logic of authentication and authorization).
